Applications/Spire/Source/Ui/DropDownList.cpp

```diff
@@ -0,0 +1,255 @@
+    case Qt::Key_Down:
+      if(isVisible()) {
+        activate_next();
+      }
+      break;
+    case Qt::Key_Enter:
+      if(m_highlight_index) {
+        on_item_selected(get_widget(*m_highlight_index)->get_value(),
+          *m_highlight_index);
+      }
+      break;
+    case Qt::Key_Escape:
+      hide();
+      break;
+    case Qt::Key_Up:
+      if(isVisible()) {
+        activate_previous();
+      }
+      break;
+  }
+  DropDownWindow::keyPressEvent(event);
+}
+
+connection DropDownList::connect_activated_signal(
+    const ActivatedSignal::slot_type& slot) const {
+  return m_activated_signal.connect(slot);
+}
+
+connection DropDownList::connect_highlighted_signal(
+    const HighlightedSignal::slot_type& slot) const {
+  return m_highlighted_signal.connect(slot);
+}
+
+connection DropDownList::connect_index_selected_signal(
+    const IndexSelectedSignal::slot_type& slot) const {
+  return m_index_selected_signal.connect(slot);
+}
+
+connection DropDownList::connect_value_selected_signal(
+    const ValueSelectedSignal::slot_type& slot) const {
+  return m_value_selected_signal.connect(slot);
+}
+
+QVariant DropDownList::get_value(int index) {
+  if(m_layout->count() - 1 < index) {
+    return QVariant();
+  }
+  return get_widget(index)->get_value();
+}
+
+void DropDownList::insert_item(DropDownItem* item) {
+  m_item_selected_connections.push_back(item->connect_selected_signal(
+    [=] (auto value) {
+      on_item_selected(std::move(value), item);
+    }));
+  m_layout->insertWidget(m_layout->count(), item);
+  update_height();
+}
+
+int DropDownList::item_count() const {
+  return m_layout->count();
+}
+
+void DropDownList::remove_item(int index) {
+  if(index > m_layout->count() - 1) {
+    return;
+  }
+  m_item_selected_connections.erase(
+    m_item_selected_connections.begin() + index);
+  auto layout_item = m_layout->takeAt(index);
+  delete layout_item->widget();
+  delete layout_item;
+  update_height();
+}
+
+bool DropDownList::set_highlight(const QString& text) {
+  for(auto i = 0; i < item_count(); ++i) {
+    if(m_item_delegate.displayText(get_value(i)).startsWith(text,
+        Qt::CaseInsensitive)) {
+      set_highlight(i);
+      return true;
+    }
+  }
+  return false;
+}
+
+void DropDownList::set_items(std::vector<DropDownItem*> items) {
+  while(auto item = m_layout->takeAt(0)) {
+    delete item->widget();
+    delete item;
+  }
+  m_item_selected_connections.clear();
+  for(auto i = std::size_t(0); i < items.size(); ++i) {
+    m_layout->addWidget(items[i]);
+    m_item_selected_connections.push_back(items[i]->connect_selected_signal(
+      [=, item = items[i]] (auto value) {
+        on_item_selected(std::move(value), item);
+      }));
+  }
+  if(m_layout->count() > 0) {
+    update_height();
+    static_cast<DropDownItem*>(m_layout->itemAt(0)->widget())->set_highlight();
+    m_highlight_index = 0;
+  } else {
+    hide();
+  }
+}
+
+DropDownItem* DropDownList::get_widget(int index) {
```

Comment:
Use a symmetric naming convention. If you use ```set_items``` as the setter, then use ```get_item``` as the getter.